In [34]:
import dash
from dash import Dash, dcc, html
from dash.dependencies import Input, Output
from dash.exceptions import PreventUpdate
from jupyter_dash import JupyterDash
import pandas as pd
import plotly.graph_objects as go
import plotly.express as px
import seaborn as sns
import pandas as pd
import numpy as np
import sys

np.set_printoptions(threshold=sys.maxsize)


In [35]:
data = pd.read_csv("NY-House-Dataset.csv")
data.head(100)

BROKERTITLE                TYPE  \
0         Brokered by Douglas Elliman  -111 Fifth Ave      Condo for sale   
1                                 Brokered by Serhant      Condo for sale   
2                              Brokered by Sowae Corp      House for sale   
3                                 Brokered by COMPASS      Condo for sale   
4   Brokered by Sotheby's International Realty - E...  Townhouse for sale   
..                                                ...                 ...   
95                 Brokered by STROFFOLINO REALTY LLC      Co-op for sale   
96     Brokered by Douglas Elliman - 3544 Johnson Ave      Co-op for sale   
97                             Brokered by Sowae Corp      House for sale   
98          Brokered by Americas Real Estate Advisors      House for sale   
99      Brokered by Douglas Elliman - 575 Madison Ave      House for sale   

        PRICE  BEDS       BATH  PROPERTYSQFT  \
0      315000     2   2.000000   1400.000000   
1   195000000     7  10.000000  17545.000000   
2      260000     4   2.000000   2015.000000   
3       69000     3   1.000000    445.000000   
4    55000000     7   2.373861  14175.000000   
..        ...   ...        ...           ...   
95     539000     2   2.000000   1100.000000   
96    1295000     4   2.000000   1995.000000   
97     245000     3   1.000000   2184.207862   
98     400000     7   3.000000   2385.000000   
99   55000000     8   8.000000  12000.000000   

                                              ADDRESS  \
0                                2 E 55th St Unit 803   
1   Central Park Tower Penthouse-217 W 57th New Yo...   
2                                    620 Sinclair Ave   
3                             2 E 55th St Unit 908W33   
4                                         5 E 64th St   
..                                                ...   
95                               9425 Shore Rd Apt 4F   
96                           2501 Palisade Ave Apt H2   
97                                   2942 Lurting Ave   
98                                     340 E 234th St   
99                                    25 Riverside Dr   

                      STATE  \
0        New York, NY 10022   
1        New York, NY 10019   
2   Staten Island, NY 10312   
3       Manhattan, NY 10022   
4        New York, NY 10065   
..                      ...   
95       Brooklyn, NY 11209   
96          Bronx, NY 10463   
97          Bronx, NY 10469   
98          Bronx, NY 10470   
99       New York, NY 10023   

                                         MAIN_ADDRESS  \
0              2 E 55th St Unit 803New York, NY 10022   
1   Central Park Tower Penthouse-217 W 57th New Yo...   
2             620 Sinclair AveStaten Island, NY 10312   
3          2 E 55th St Unit 908W33Manhattan, NY 10022   
4                       5 E 64th StNew York, NY 10065   
..                                                ...   
95             9425 Shore Rd Apt 4FBrooklyn, NY 11209   
96            2501 Palisade Ave Apt H2Bronx, NY 10463   
97                    2942 Lurting AveBronx, NY 10469   
98                      340 E 234th StBronx, NY 10470   
99                  25 Riverside DrNew York, NY 10023   

   ADMINISTRATIVE_AREA_LEVEL_2      LOCALITY      SUBLOCALITY  \
0              New York County      New York        Manhattan   
1                United States      New York  New York County   
2                United States      New York  Richmond County   
3                United States      New York  New York County   
4                United States      New York  New York County   
..                         ...           ...              ...   
95                    New York  Kings County         Brooklyn   
96                    New York  Bronx County        The Bronx   
97               United States      New York     Bronx County   
98               United States      New York     Bronx County   
99               United States      New York  New York County   

         STREET_

In [36]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4801 entries, 0 to 4800
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   BROKERTITLE                  4801 non-null   object 
 1   TYPE                         4801 non-null   object 
 2   PRICE                        4801 non-null   int64  
 3   BEDS                         4801 non-null   int64  
 4   BATH                         4801 non-null   float64
 5   PROPERTYSQFT                 4801 non-null   float64
 6   ADDRESS                      4801 non-null   object 
 7   STATE                        4801 non-null   object 
 8   MAIN_ADDRESS                 4801 non-null   object 
 9   ADMINISTRATIVE_AREA_LEVEL_2  4801 non-null   object 
 10  LOCALITY                     4801 non-null   object 
 11  SUBLOCALITY                  4801 non-null   object 
 12  STREET_NAME                  4801 non-null   object 
 13  LONG_NAME         

* BROKERTITLE: Title of the broker
* TYPE: Type of the house
* PRICE: Price of the house
* BEDS: Number of bedrooms
* BATH: Number of bathrooms
* PROPERTYSQFT: Square footage of the property
* ADDRESS: Full address of the house
* STATE: State of the house
* MAIN_ADDRESS: Main address information
* ADMINISTRATIVE_AREA_LEVEL_2: Administrative area level 2 information
* LOCALITY: Locality information
* SUBLOCALITY: Sublocality information
* STREET_NAME: Street name
* LONG_NAME: Long name
* FORMATTED_ADDRESS: Formatted address
* LATITUDE: Latitude coordinate of the house
* LONGITUDE: Longitude coordinate of the house

In [37]:
data["BROKERTITLE"].unique()
data["TYPE"].unique()
data["PRICE"].unique()
data["BEDS"].unique()
data["BATH"].unique()
# data["PROPERTYSQFT"].unique()
# data["ADDRESS"].unique()
# data["STATE"].unique()
# data["MAIN_ADDRESS"].unique()
# data["ADMINISTRATIVE_AREA_LEVEL_2"].unique()
# data["LOCALITY"].unique()
# data["SUBLOCALITY"].unique()
# data["STREET_NAME"].unique()
# data["LONG_NAME"].unique()
# data["FORMATTED_ADDRESS"].unique()
# data["LATITUDE"].unique()
# data["LONGITUDE"].unique()


array([ 2.        , 10.        ,  1.        ,  2.37386086, 16.        ,
        3.        ,  4.        ,  6.        ,  8.        ,  5.        ,
        9.        ,  7.        , 32.        , 13.        , 50.        ,
       20.        , 11.        , 12.        , 24.        , 43.        ,
        0.        , 17.        ])

In [38]:
datax = data.copy()

### column BROKERTITLE:

1. delete all 'Brokered by'
2. delete leading and trailing whitespaces


In [39]:
datax['BROKERTITLE1'] = datax['BROKERTITLE'].str.replace('Brokered by ', "").str.strip()


### column TYPE:

https://www.manhattanmiami.com/resources/home-buying-guide/co-op-vs-condo

### columns BATH, PROPERTYSQFT:

strange value = 2.37386086  ??? rounded!
in csv = 2.3738608579684373


In [40]:
datax.loc[datax['BATH'] == 2.3738608579684373, :]


BROKERTITLE  \
4     Brokered by Sotheby's International Realty - E...   
11                              Brokered by Re/Max Edge   
27                  Brokered by Century 21 Realty First   
68                            Brokered by Ashford Homes   
69    Brokered by Sotheby's International Realty - E...   
...                                                 ...   
4776                   Brokered by CENTURY 21 KR REALTY   
4780                                           NoBroker   
4783                  Brokered by Silverside Realty LLC   
4788           Brokered by J Shayovitz Real Estate Corp   
4790                                Brokered by Serhant   

                            TYPE     PRICE  BEDS      BATH  PROPERTYSQFT  \
4             Townhouse for sale  55000000     7  2.373861  14175.000000   
11            Townhouse for sale    689000     3  2.373861   1162.000000   
27    Multi-family home for sale   2250000    12  2.373861   5670.000000   
68    Multi-family home for sale   1489000     6  2.373861   3780.000000   
69            Townhouse for sale  65000000     3  2.373861  15200.000000   
...                          ...       ...   ...       ...           ...   
4776                     Pending    555000     3  2.373861   2184.207862   
4780  Multi-family home for sale   2000000     8  2.373861   2184.207862   
4783  Multi-family home for sale    799000     6  2.373861   2000.000000   
4788              Co-op for sale   4750000     3  2.373861   2969.000000   
4790              Condo for sale    789000     3  2.373861    800.000000   

                      ADDRESS                    STATE  \
4                 5 E 64th St       New York, NY 10065   
11           456 Van Name Ave  Staten Island, NY 10303   
27               2361 81st St       Brooklyn, NY 11214   
68              205 Wiman Ave  Staten Island, NY 10308   
69                4 E 79th St       New York, NY 10075   
...                       ...                      ...   
4776         780 Stafford Ave  Staten Island, NY 10309   
4780             2324 60th St       Brooklyn, NY 11204   
4783       4568 Carpenter Ave          Bronx, NY 10470   
4788    1175 Park Ave Unit 1C       New York, NY 10128   
4790  109 Lafayette St Rm 502       New York, NY 10013   

                                   MAIN_ADDRESS ADMINISTRATIVE_AREA_LEVEL_2  \
4                 5 E 64th StNew York, NY 10065               United States   
11      456 Van Name AveStaten Island, NY 10303               United States   
27               2361 81st StBrooklyn, NY 11214                       11214   
68         205 Wiman AveStaten Island, NY 10308               United States   
69                4 E 79th StNew York, NY 10075               United States   
...                                         ...                         ...   
4776    780 Stafford AveStaten Island, NY 10309               United States   
4780             2324 60th StBrooklyn, NY 11204               United States   
4783          4568 Carpenter AveBronx, NY 10470               United States   
4788    1175 Park Ave Unit 1CNew York, NY 10128                    New York   
4790  109 Lafayette St Rm 502New York, NY 10013                    New York   

             LOCALITY      SUBLOCALITY    STREET_NAME         LONG_NAME  \
4            New York  New York County       New York  East 64th Street   
11           New York  Richmond County  Staten Island   Van Name Avenue   
27      United States         New York   Kings County       81st Street   
68           New York  Richmond County  Staten Island      Wiman Avenue   
69           New York  New York County       New York  East 79th Street   
...               ...              ...            ...               ...   
4776         New York  Richmond County  Staten Island   Stafford Avenue   
4780         New York     Kings County       Brooklyn       60th Street   
4783         New York     Bronx County      The Bronx  Carpenter Avenue   
4788  New York County     

In [41]:
datax['BATH1'] = datax['BATH'].astype('int64')

In [42]:
datax['BATH1'].unique()

array([ 2, 10,  1, 16,  3,  4,  6,  8,  5,  9,  7, 32, 13, 50, 20, 11, 12,
       24, 43,  0, 17])

In [43]:
datax.loc[datax['PROPERTYSQFT'] == 2184.207862, :]

BROKERTITLE            TYPE  \
6         Brokered by Douglas Elliman - 575 Madison Ave  Condo for sale   
12                                  Brokered by COMPASS  Co-op for sale   
13           Brokered by Jamie & Connie Real Estate Grp  Co-op for sale   
14                Brokered by Corcoran Chelsea/Flatiron  Co-op for sale   
26    Brokered by Nest Seekers International, Long I...  Co-op for sale   
...                                                 ...             ...   
4791                     Brokered by Corcoran East Side  Co-op for sale   
4793                       Brokered by Engel & V�������  Co-op for sale   
4796                                Brokered by COMPASS  Co-op for sale   
4797                    Brokered by Mjr Real Estate Llc  Co-op for sale   
4798      Brokered by Douglas Elliman - 575 Madison Ave  Co-op for sale   

        PRICE  BEDS  BATH  PROPERTYSQFT                    ADDRESS  \
6      899500     2   2.0   2184.207862     157 W 126th St Unit 1B   
12     259000     3   1.0   2184.207862      34-41 85th St Unit 1D   
13     430000     2   2.0   2184.207862   91-15 Lamont Ave Unit 6D   
14     895000     3   1.0   2184.207862          61 Jane St Apt 6N   
26     325000     1   1.0   2184.207862      35-45 81st St Unit E1   
...       ...   ...   ...           ...                        ...   
4791   370000     3   1.0   2184.207862       342 E 53rd St Apt 6C   
4793  1950000     2   4.0   2184.207862      700 Park Ave Unit 20A   
4796   599000     1   1.0   2184.207862       222 E 80th St Apt 3A   
4797   245000     1   1.0   2184.207862        97-40 62 Dr Unit Lg   
4798  1275000     1   1.0   2184.207862  427 W 21st St Unit Garden   

                          STATE  \
6            New York, NY 10027   
12    Jackson Heights, NY 11372   
13           Elmhurst, NY 11373   
14           New York, NY 10014   
26             Queens, NY 11372   
...                         ...   
4791         New York, NY 10022   
4793         New York, NY 10021   
4796        Manhattan, NY 10075   
4797        Rego Park, NY 11374   
4798         New York, NY 10011   

                                        MAIN_ADDRESS  \
6           157 W 126th St Unit 1BNew York, NY 10027   
12    34-41 85th St Unit 1DJackson Heights, NY 11372   
13        91-15 Lamont Ave Unit 6DElmhurst, NY 11373   
14               61 Jane St Apt 6NNew York, NY 10014   
26             35-45 81st St Unit E1Queens, NY 11372   
...                                              ...   
4791          342 E 53rd St Apt 6CNew York, NY 10022   
4793         700 Park Ave Unit 20ANew York, NY 10021   
4796         222 E 80th St Apt 3AManhattan, NY 10075   
4797          97-40 62 Dr Unit LgRego Park, NY 11374   
4798     427 W 21st St Unit GardenNew York, NY 10011   

     ADMINISTRATIVE_AREA_LEVEL_2         LOCALITY      SUBLOCALITY  \
6                       New York  New York County         New York   
12                      New York    Queens County           Queens   
13                      New York    Queens County           Queens   
14                      New York  New York County         New York   
26                      New York    Queens County           Queens   
...                          ...              ...              ...   
4791                    New York  New York County         New York   
4793                    New York  New York County         New York   
4796                    New York  New York County         New York   
4797               United States         New York    Queens County   
4798               United States         New York  New York County   

          STREET_NAME         LONG_NAME  \
6           Manhattan               157   
12           Flushing             34-41   
13           Elmhurst             91-15   
14          Manhattan                61   
26    Jackson Heights             35-45   
...               ...               ...   
4791        Manhattan               342   
4793        Manhattan  

In [44]:
datax['PROPERTYSQFT1'] = datax['PROPERTYSQFT'].astype('int64')


### column ADDRESS, STATE, MAIN_ADDRESS:

In [45]:
datax['ADDRESS1'] = datax['ADDRESS'].str.strip()
datax['MAIN_ADDRESS1'] = datax['MAIN_ADDRESS'].str.strip()


In [46]:
datax['STATE1'] = 'New York'

### columns of location hierarchy:


* ZIP - using regex formula to extract zip code from 'FORMATTED_ADDRESS' column

* COUNTY, CITY - using zipcodes module to extract counties/cities from it (using for loop to change each result for each zip code from list of dictionaries to a dictionary and make a column by list.append(dict.get('key')))

* STREET_NAME - using column FORMATTED_ADDRESS to extract names of streets (using package 'usaddress', I have also tried to use geopy.geocoders - Nominatim based on lat and long, but it is inapplicable due to amount of rows)

* delete ['ADMINISTRATIVE_AREA_LEVEL_2', 'LOCALITY', 'SUBLOCALITY'] = inconsistent, and replace them by new columns 

United States is structured as follows: Admin 0 - Nation, Admin 1 States along with Federal District (DC), Admin 2 - Counties & Independent Cities, etc.

In [ ]:
! pip install zipcodes


In [48]:
import zipcodes
print(zipcodes.matching('11214'))

[{'zip_code': '11214', 'zip_code_type': 'STANDARD', 'active': True, 'city': 'Brooklyn', 'acceptable_cities': [], 'unacceptable_cities': [], 'state': 'NY', 'county': 'Kings County', 'timezone': 'America/New_York', 'area_codes': ['718'], 'world_region': 'NA', 'country': 'US', 'lat': '40.5977', 'long': '-74.0010'}]


In [49]:
omg = zipcodes.matching('11214')


In [50]:
type(omg) # omg is a list of dictionaries

list

In [51]:
import re
datax['ZIP'] = datax['FORMATTED_ADDRESS'].str.extract('(?<=NY )(.*?)(?=, USA)', expand=True)


In [52]:
# datax['ZIP'].unique()
datax.loc[datax['ZIP'] == '113725105', 'ZIP'] = '11372'

In [53]:
county = []

for value in datax["ZIP"]:
    dict = {}
    for i in zipcodes.matching(value):
        dict.update(i)
    county.append(dict.get('county'))

datax['COUNTY'] = county

In [54]:
city = []

for value in datax["ZIP"]:
    dict1 = {}
    for i in zipcodes.matching(value):
        dict1.update(i)
    city.append(dict1.get('city'))

datax['CITY'] = city

In [71]:
datax.drop(['ADMINISTRATIVE_AREA_LEVEL_2', 'LOCALITY', 'SUBLOCALITY'], inplace=True, axis=1)
datax.drop(['BROKERTITLE', 'BATH', 'PROPERTYSQFT', 'ADDRESS', 'STATE', 'MAIN_ADDRESS'], inplace=True, axis=1)

In [56]:
# ! pip install geopy

In [57]:
# from geopy.geocoders import Nominatim
# from pprint import pprint

# # Instantiate a new Nominatim client
# geolocator = Nominatim(user_agent="NY_address")

# # Get location raw data from the user

# location = geolocator.reverse("40.761255, -73.974483")
# print(location.address)

In [58]:
# datax['COORDINATES'] = datax['LATITUDE'].astype(str) + ', ' + datax['LONGITUDE'].astype(str)

In [59]:
# addressLL = []

# for value in datax['COORDINATES']:
#     geolocator = Nominatim(user_agent="NY_address")
#     location = geolocator.reverse(value)
#     addressLL.append(location.address)

# datax.drop('COORDINATES', inplace=True, axis=1)

In [ ]:
! pip install usaddress

In [61]:
import usaddress

In [62]:
# addr='217 W 57th St, New York, NY 10019, USA'
# test = usaddress.parse(addr)
# print(test)
# # test2 = dict(test)
# # print(test2)
# result1 = [item for item in test if item[1] == 'StreetName']
# result2 = [item for item in test if item[1] == 'StreetNamePostType']
# print(result1[0][0]  + ' ' + result2[0][0])

In [63]:
# street_name = []

# for value in datax['ADDRESS1']:
#     try:
#         res1 = [item for item in usaddress.parse(value) if item[1] == 'StreetName']
#         res2 = [item for item in usaddress.parse(value) if item[1] == 'StreetNamePostType']
#         street_name.append(res1[0][0] + ' ' + res2[0][0])

#     except IndexError as e:
#             street_name.append('NaN')

# datax['STR_NAME'] = street_name

# datax.drop('STREET_NAME', inplace=True, axis=1)
# datax.drop('LONG_NAME', inplace=True, axis=1)

In [64]:
# datax['STR_NAME2'] = datax['STR_NAME'].str.replace(r"\W", " ", regex=True).str.strip()

In [72]:
test = []
for value in datax['FORMATTED_ADDRESS']:
    try:
        res1 = [item for item in usaddress.parse(value) if item[1] == 'StreetName']
        res2 = [item for item in usaddress.parse(value) if item[1] == 'StreetNamePostType']
        test.append(res1[0][0] + ' ' + res2[0][0])
    except IndexError:
        if 'Avenue'in value:
            test.append(re.findall(r"(Avenue\s.)", value)[0])
        elif 'Broadway' in value:
            test.append('Broadway')
        elif '55th St' in value:
            test.append('55th St') 
        elif 'Central Park' in value:
            test.append(re.findall(r"(Central Park\s.)", value)[0])
        elif 'Prospect Park W' in value:
            test.append('Prospect Park W')
        elif 'Tudor City' in value:
            test.append('Tudor City Pl')
        elif 'South St' in value:
            test.append('South St')
        elif '164th Dr' in value:
            test.append('164th Dr')
        elif 'Ocean Promenade' in value:
            test.append('Ocean Promenade')
        elif 'Fifth Ave' in value:
            test.append('Fifth Ave')
        elif 'Bowery' in value:
            test.append('Bowery')
        elif 'Grand Concourse' in value:
            test.append('Grand Concourse')
        elif 'Morris Park W' in value:            
            test.append('Morris Park W')
        elif 'First Ave' in value:
            test.append('First Ave')
        elif 'Court Sq W' in value:
            test.append('Court Sq W')
        elif 'Northside Piers' in value:
            test.append('Northside Piers')
        elif 'Alden Park' in value:
            test.append('Alden Park')
        elif 'Hudson Yards' in value:
            test.append('Hudson Yards')
        elif 'Ave W' in value:
            test.append('Avenue W')
        elif 'Gramercy Park S' in value:
            test.append('Gramercy Park S')
        elif 'Ave V' in value:
            test.append('Avenue V')
        elif 'Ave Y' in value:
            test.append('Avenue Y')
        elif 'Ave. S' in value:
            test.append('Avenue S')
        elif '5th Ave' in value:
            test.append('5th Avenue')
        elif 'Ocean Prom' in value:
            test.append('Ocean Promenade')
        elif '1st Ave' in value:
            test.append('1st Avenue')
        elif 'Ct Square W' in value:
            test.append('Court Square W')
        elif 'Gramercy Pk S' in value:
            test.append('Gramercy Park S')
        else:
            test.append('NaN')

datax['TEST'] = test



In [73]:
datax['STR_NAME'] = datax['TEST'].str.replace(r"\W", " ", regex=True).str.strip()

In [74]:
datax.drop('TEST', inplace=True, axis=1)

In [75]:
datax.loc[datax['STR_NAME'] == 'NaN', :]

Empty DataFrame
Columns: [TYPE, PRICE, BEDS, STREET_NAME, LONG_NAME, FORMATTED_ADDRESS, LATITUDE, LONGITUDE, BROKERTITLE1, BATH1, PROPERTYSQFT1, ADDRESS1, MAIN_ADDRESS1, STATE1, ZIP, COUNTY, CITY, STR_NAME]
Index: []

### Change column names and sort them properly, delete some unnecessary address columns:

In [82]:
# datax.drop(['ADDRESS1', 'MAIN_ADDRESS1'], axis=1, inplace=True)

# datax.columns.tolist()

# datax.rename(columns = {'BROKERTITLE1': 'BROKERTITLE',
#                         'BATH1': 'BATH',
#                         'PROPERTYSQFT1': 'PROPERTYSQFT',
#                         'STATE1': 'STATE'}, 
#                         inplace = True)

new_cols = ['BROKERTITLE', 
            'TYPE',
            'PRICE',
            'BEDS',
            'BATH',
            'PROPERTYSQFT',
            'FORMATTED_ADDRESS',
            'STATE',
            'ZIP',
            'COUNTY',
            'CITY',
            'STR_NAME',
            'LATITUDE',
            'LONGITUDE']

data_cl = datax[new_cols]


In [83]:
data_cl.head()

BROKERTITLE                TYPE  \
0                    Douglas Elliman  -111 Fifth Ave      Condo for sale   
1                                            Serhant      Condo for sale   
2                                         Sowae Corp      House for sale   
3                                            COMPASS      Condo for sale   
4  Sotheby's International Realty - East Side Man...  Townhouse for sale   

       PRICE  BEDS  BATH  PROPERTYSQFT  \
0     315000     2     2          1400   
1  195000000     7    10         17545   
2     260000     4     2          2015   
3      69000     3     1           445   
4   55000000     7     2         14175   

                                   FORMATTED_ADDRESS     STATE    ZIP  \
0  Regis Residence, 2 E 55th St #803, New York, N...  New York  10022   
1             217 W 57th St, New York, NY 10019, USA  New York  10019   
2     620 Sinclair Ave, Staten Island, NY 10312, USA  New York  10312   
3               2 E 55th St, New York, NY 10022, USA  New York  10022   
4               5 E 64th St, New York, NY 10065, USA  New York  10065   

            COUNTY           CITY      STR_NAME   LATITUDE  LONGITUDE  
0  New York County       New York       55th St  40.761255 -73.974483  
1  New York County       New York       57th St  40.766393 -73.980991  
2  Richmond County  Staten Island  Sinclair Ave  40.541805 -74.196109  
3  New York County       New York       55th St  40.761398 -73.974613  
4  New York County       New York       64th St  40.767224 -73.969856

In [ ]:
data_cl.to_csv('NY_housing_cleared.csv')